src/test/StatedWorkflow.test.js

```diff
@@ -1027,4 +1032,123 @@ test("subscribePulsar with pulsarMock client", async () => {
+    } catch (e) {
+        if (e.code !== 'ENOENT') {
+            throw e;
+        }
+    }
+
+    await tp.initialize();
+    console.log("Initialized stated workflow template...");
+
+    let snapshot;
+
+    // Wait for snapshot to capture field simulateFailure set to true
+    while (true) {
+            // read snapshot and check if output.simulateFailure is set to true
+            try {
+                const snapshotContent = fs.readFileSync(snapshotRelativePath, 'utf8');
+                snapshot = JSON.parse(snapshotContent);
+            } catch (e) {
+                if (e.code === 'ENOENT' || e.message === 'Unexpected end of JSON input') {
+                    console.log(`Snapshot is not ready yet: ${snapshotRelativePath}, waiting...`)
+                    await new Promise(resolve => setTimeout(resolve, 1000)); // Poll every 1s
+                    continue; // waiting for snapshot to be made
+                }
+                console.error(`Error checking snapshot residents: ${e.message}`);
+                throw e;
+            }
+            console.log(`Snapshot has ${snapshot.output.rebels.length} rebels, waiting for output.simulateFailure set to true...`);
+            if (snapshot.output.simulateFailure === false) {
+                // by the test design, failure occurs when we're processing second rebel 'han'. 'luke' should be saved and
+                // acknowledged.
+                expect(snapshot.output.rebels.length).toEqual(1);
+                expect(StatedREPL.stringify(snapshot.output.rebels)).toEqual(StatedREPL.stringify([
+                    {"name": "Luke Skywalker", "url": "https://swapi.dev/api/planets/1/"}]));
+                expect(StatedREPL.stringify(snapshot.output.subscribeParams.acks)).toEqual(StatedREPL.stringify(["luke"]));
+                // the output should also have a log for 'han' invocationId complete for fetchRebel
+                expect (snapshot.output.fetchRebel.log.han.end).toBeDefined();
+
+                // the second step saveRebel is designed to hang on 'han' and not to complete.
+                // ... it should have a start log
+                expect (snapshot.output.saveRebel.log.han.start).toBeDefined();
+                // ... but no end log.
+                expect (snapshot.output.saveRebel.log.han.end).toBeUndefined();
+                break;
+            }
+        await new Promise(resolve => setTimeout(resolve, 1000)); // Poll every 1s
+    };
+
+    // Kill the wf
+    console.log("Stopping stated workflow with ${tp.output.rebels.length} rebels saved in the output...");
+    await sw.close();
+    console.log(`Stopped stated workflow`);
+
+    console.log(`Restore from a snapshot...`);
+    sw = await StatedWorkflow.newWorkflow(template);
+    tp = sw.templateProcessor;
+    await TemplateProcessor.prepareSnapshotInPlace(snapshot);
+    await tp.initialize(snapshot.template, '/', snapshot.output);
+
+    while (tp.output.rebels.length !== 3) {
```

Comment:
why not use a dataChangeListener?

Reply:
it just follows a pattern, where we poll from snapshot. We could though in this case.